File: 15_Segment-Tree/AN/BOJ2357.py
import sys
input = sys.stdin.readline
INF = sys.maxsize

def min_init(node, start, end):
    if start == end:
        min_tree[node] = array[start]
    else:
        mid = (start + end) // 2
        min_tree[node] = min(min_init(node*2, start, mid), min_init(node*2+1, mid+1, end))
    return min_tree[node]

def max_init(node, start, end):
    if start == end:
        max_tree[node] = array[start]
    else:
        mid = (start + end) // 2
        max_tree[node] = max(max_init(node*2, start, mid), max_init(node*2+1, mid+1, end))
    return max_tree[node]

def subMin(node, start, end, left, right):
    if left > end or right < start:
        return INF
    if left <= start and right >= end:
        return min_tree[node]

    mid = (start + end) // 2
    return min(subMin(node*2, start, mid, left, right), subMin(node*2+1, mid+1, end, left, right))

def subMax(node, start, end, left, right):
    if left > end or right < start:
        return -1
    if left <= start and right >= end:
        return max_tree[node]

    mid = (start + end) // 2
    return max(subMax(node*2, start, mid, left, right), subMax(node*2+1, mid+1, end, left, right))


min_tree = [INF for _ in range(262144)]
max_tree = [0 for _ in range(262144)]
array = []

N,M = map(int, input().split())

for _ in range(N):
    array.append(int(input()))

min_init(1,0,N-1)
max_init(1,0,N-1)

for _ in range(M):
    a,b = map(int,input().split())
    print(subMin(1,0,N-1,a-1,b-1), subMax(1,0,N-1,a-1,b-1))

File: 15_Segment-Tree/SONG/BOJ2357.java
package segment_tree;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ2357 {
	static int[] arr, minTree, maxTree;

	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());

		arr = new int[N + 1];
		for (int i = 1; i <= N; i++) {
			arr[i] = Integer.parseInt(br.readLine());
		}

		minTree = new int[N * 4];
		maxTree = new int[N * 4];

		minInit(1, N, 1);
		maxInit(1, N, 1);

		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());

			int left = Integer.parseInt(st.nextToken());
			int right = Integer.parseInt(st.nextToken());

			sb.append(minFind(1, N, 1, left, right) + " " + maxFind(1, N, 1, left, right) + "\n");
		}

		System.out.println(sb.toString());
	}

	public static int minInit(int start, int end, int node) {
		if (start == end) {
			return minTree[node] = arr[start];
		}

		int mid = (start + end) / 2;
		return minTree[node] = Math.min(minInit(start, mid, node * 2), minInit(mid + 1, end, node * 2 + 1));
	}

	public static int maxInit(int start, int end, int node) {
		if (start == end) {
			return maxTree[node] = arr[start];
		}

		int mid = (start + end) / 2;
		return maxTree[node] = Math.max(maxInit(start, mid, node * 2), maxInit(mid + 1, end, node * 2 + 1));
	}

	public static int minFind(int start, int end, int node, int left, int right) {
		if (right < start || end < left) {
			return Integer.MAX_VALUE;
		}

		if (left <= start && end <= right) {
			return minTree[node];
		}

		int mid = (start + end) / 2;

		return Math.min(minFind(start, mid, node * 2, left, right), minFind(mid + 1, end, node * 2 + 1, left, right));
	}

	public static int maxFind(int start, int end, int node, int left, int right) {
		if (right < start || end < left) {
			return Integer.MIN_VALUE;
		}

		if (left <= start && end <= right) {
			return maxTree[node];
		}

		int mid = (start + end) / 2;

		return Math.max(maxFind(start, mid, node * 2, left, right), maxFind(mid + 1, end, node * 2 + 1, left, right));
	}

}